### ISSUE TITLE
[Feature] Support when the template content is changed, the update of a single component only affects its own k8s resources

### ISSUE THREAD

**Issue by zxbyoyoyo** [kubevela/kubevela#4702]
**Describe the bug**

Scenario description:

now we will put multiple microservice components in an Application, each release is a single microservice component release
YAML example:
````
app1
  component:
    - type: micro-service
      name: s1
      props: s1 parameter
    - type: micro-service
      name: s2
      props: s2 parameter
````
When the content of the micro-service component template is changed, the parameters of s2 are modified at this time, and when the Application update is triggered, the k8s resources associated with the s1 and s2 components will be updated. I want that s1 will not be updated; 
To put it simply, i want the release of s2 service does not affect s1 service

**To Reproduce**

1. The YAML files of ComponentI used. container only one env
vela def apply microservice.yaml
```
microservice: {
	type: "component"
	annotations: {}
	labels: {}
	description: "Describes long-running, scalable, containerized services that have a stable network endpoint to receive external network traffic from customers."
	attributes: {
		workload: {
			definition: {
				apiVersion: "apps/v1"
				kind:       "Deployment"
			}
			type: "deployments.apps"
		}
	}
}
template: {

	output: {
		apiVersion: "apps/v1"
		kind:       "Deployment"
		spec: {
			selector: matchLabels: {
				"app": context.name
			}
			template: {
				metadata: {
					labels: {
						"app": context.name
					}
					annotations: {
						"version": parameter.version
					}
				}

				spec: {
					containers: [{
						name:  context.name
						image: parameter.image
						ports: [{
							containerPort: parameter.port
						}]
						env: [{
							name:  "default_env01"
							value: "default_env01"
						}]

					}]

				}
			}
		}
	}

	parameter: {
		image:    string
		port?:    int
		version?: string
	}

}

```
2. The YAML file of Application I applied.
``` shell
cat <<EOF | kubectl apply -f -
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: microservice-app
spec:
  components:
    - name: s1
      type: microservice
      properties:
        image: crccheck/hello-world
        port: 8080
        version: "1"
    - name: s2
      type: microservice
      properties:
        image: crccheck/hello-world
        port: 8080
        version: "1"
EOF
```
3. See Output
```
 kubectl get pods                                                              
NAME                                   READY   STATUS    RESTARTS   AGE
s1-7cbdd6bdf7-bhpz2                    1/1     Running   0          6m
s2-687cfd5c59-9zmfb                    1/1     Running   0          6m

```
4. Modify the component template ,e.g. add an env

```
microservice: {
	type: "component"
	annotations: {}
	labels: {}
	description: "Describes long-running, scalable, containerized services that have a stable network endpoint to receive external network traffic from customers."
	attributes: {
		workload: {
			definition: {
				apiVersion: "apps/v1"
				kind:       "Deployment"
			}
			type: "deployments.apps"
		}
	}
}
template: {

	output: {
		apiVersion: "apps/v1"
		kind:       "Deployment"
		spec: {
			selector: matchLabels: {
				"app": context.name
			}
			template: {
				metadata: {
					labels: {
						"app": context.name
					}
					annotations: {
						"version": parameter.version
					}
				}

				spec: {
					containers: [{
						name:  context.name
						image: parameter.image
						ports: [{
							containerPort: parameter.port
						}]
						env: [{
							name:  "default_env01"
							value: "default_env01"
						},{
							name:  "default_env02"
							value: "default_env02"
						}]

					}]

				}
			}
		}
	}

	parameter: {
		image:    string
		port?:    int
		version?: string
	}

}

```
5.Update the properties of s2
```shell
cat <<EOF | kubectl apply -f -
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: microservice-app
spec:
  components:
    - name: s1
      type: microservice
      properties:
        image: crccheck/hello-world
        port: 8080
        version: "1"
    - name: s2
      type: microservice
      properties:
        image: crccheck/hello-world
        port: 8080
        version: "2"
EOF
```

6. See output，the k8s resources of both components have been updated
```
s1-7866b94744-bn8bf                    0/1     ContainerCreating   0          5s
s1-7cbdd6bdf7-bhpz2                    1/1     Running             0          10m
s2-687cfd5c59-9zmfb                    1/1     Running             0          10m
s2-7c95b6c49f-vm9vr                    0/1     ContainerCreating   0          4s

```

**KubeVela Version**

Core Version: 1.5.0

**Cluster information**

**Additional context**

**Comment by wonderflow** [comment]
In your example case, why don't use two apps for this work?

**Comment by zxbyoyoyo** [comment]
> In your example case, why don't use two apps for this work?

Our platform design is that an application contains multiple microservices. So we use KubeVela like this:
- Application = KubeVela Application
- Microservice = KubeVela Component
When publishing, you can publish the application, or a microservice under the application.

If KubeVela can't achieve that the components do not affect each other, we need to create a KubeVela Application for each microservice, and we also need to maintain the concept of "application" on top of the KubeVela Application, which greatly increases our management complexity.

**Comment by wonderflow** [comment]
Thanks @zxbyoyoyo 

Now I understand this case better. I have a question, will you need to update all component at once? If your issue here is implemented, it will mean any changes on the application spec won't trigger the component to upgrade if the component/trait not updated.

1. Should change the metadata of application CR affect any component to upgrade?
2. Should change policy/workflow step affect the component?

How to distinguish these cases? 


A possible solution is to add a policy for this feature request. When the policy added, only changes of spec in the component level will trigger the component to upgrade.

**Comment by zxbyoyoyo** [comment]
Thanks @wonderflow  

This is a nice solution

The Controller logic is that if the app.Spec(Components\Policies\Workflow) changes, the workflow will be re-executed, which I think is reasonable.

now:
Component template changed || Component properties changed => Update k8s underlying resources.

want:
Component properties changed && component template changed => Update k8s underlying resources.

So I need to customize a workflow step, the function is similar to apply-component, it will first compare whether the properties of the current component have changed, and then execute apply if there are changes;

The pseudo-code
```
for comp in Application.spec.components{
  if comp.properties == lastApplicationRevision["xxx"].properties {
       continue
    }else {
        apply
    }
}
```
ps : the effect is to be verified

**Comment by Somefive** [comment]
> Component template changed && Component properties changed => Update k8s underlying resources.

**Comment by zxbyoyoyo** [comment]
> > Component template changed && Component properties changed => Update k8s underlying resources.
> 
> This is weird. If I understand it correct, the component properties changes without template changed will not update underlying resources? I think you mean "Only update k8s underlying resources when component properties changed, component template change will not cause k8s underlying resources updates", it is correct?

you get it

**Comment by Somefive** [comment]
By far, the deploy of application is treated as a whole. We recommend to split your application into multiple applications if you need to handle your "components" separately. There are scenarios that component dispatch is intended even if the properties of the component is not changed.

The example you raised where multiple components use the same component definition but want them to update differently when the component definition is changed, violates the principle that your application follows the consistent desired state. If you insist that these two component needs to be handled together in one application, I support you to change the component type directly, like the following

```yaml
# before
- type: microservice
  name: component-1
  properties:
    key: value1
- type: microservice
  name: component-2
  properties:
    key: value2

# after
- type: microservice-new
  name: component-1
  properties:
    key: val1
- type: microservice
  name: component-2
  properties:
    key: value2
```

In this case, you keep the definition of `microservice` unchanged but instead, you add a new definition named `microservice-new` to represent the modified `microservice` component definition. Then change the properties of the component to trigger the application redeploy.

**Comment by zxbyoyoyo** [comment]
In this way,  we need to additionally maintain the components of the application and the version it uses

**Comment by zxbyoyoyo** [comment]
@wonderflow @Somefive 
I tried to modify the multicluster.#Deploy logic to implement the Apply operation only when the properties of a component changed, but found that it would conflict with the GC logic.
 
GC logic: Kubevela will generate a new tracking object (ReouceTracker) every time the App is updated to record the applied resources. For example, the first time ReouceTrackerV1: ["serviceA", "serviceB"], the second time only the servcieB component is updated, ReouceTrackerV2: ["serviceB"], at this time, GC will be based on the records of ReouceTrackerV2, to recover the unrecorded resource (servcieA's resource), so it doesn't work.

finally, we decided to accept @Somefive  suggestion, Thank you all~
